File: web/cockpit/src/lib/types/history-scan.ts
/**
 * TypeScript types for History Scan API
 * Matches backend Pydantic models from agent_platform/history_scan/models.py
 */

export type ScanStatus = 'not_started' | 'in_progress' | 'paused' | 'completed' | 'failed';

export interface ScanConfig {
  account_id: string;
  batch_size?: number;
  max_results?: number | null;
  query?: string;
  skip_already_processed?: boolean;
  process_attachments?: boolean;
  process_threads?: boolean;
}

export interface ScanProgress {
  scan_id: string;
  account_id: string;
  status: ScanStatus;

  // Progress counters
  total_found: number;
  processed: number;
  skipped: number;
  failed: number;

  // Classification breakdown
  classified_high: number;
  classified_medium: number;
  classified_low: number;

  // Extraction stats
  tasks_extracted: number;
  decisions_extracted: number;
  questions_extracted: number;

  // Attachment & thread stats
  attachments_downloaded: number;
  threads_summarized: number;

  // Timing
  started_at: string;
  last_updated_at: string;
  completed_at: string | null;
  estimated_completion: string | null;

  // Resume capability
  last_processed_email_id: string | null;
  next_page_token: string | null;

  // Error tracking
  error_message: string | null;
  error_details: Record<string, any> | null;
}

export interface ScanResult {
  scan_id: string;
  account_id: string;
  status: ScanStatus;

  // Summary stats
  total_processed: number;
  total_skipped: number;
  total_failed: number;

  // Time taken
  duration_seconds: number;
  emails_per_second: number;

  // Classification summary
  high_confidence: number;
  medium_confidence: number;
  low_confidence: number;

  // Extraction summary
  total_tasks: number;
  total_decisions: number;
  total_questions: number;

  // Attachment & thread summary
  total_attachments: number;
  total_threads: number;

  // Error summary
  errors: Array<Record<string, any>>;
}

// Helper to calculate progress percentage
export function calculateProgress(progress: ScanProgress): number {
  if (progress.total_found === 0) return 0;
  return Math.round((progress.processed / progress.total_found) * 100);
}

// Helper to format duration
export function formatDuration(startedAt: string, completedAt?: string | null): string {
  const start = new Date(startedAt);
  const end = completedAt ? new Date(completedAt) : new Date();
  const diffSeconds = Math.floor((end.getTime() - start.getTime()) / 1000);

  if (diffSeconds < 60) return `${diffSeconds}s`;
  if (diffSeconds < 3600) return `${Math.floor(diffSeconds / 60)}m ${diffSeconds % 60}s`;
  const hours = Math.floor(diffSeconds / 3600);
  const minutes = Math.floor((diffSeconds % 3600) / 60);
  return `${hours}h ${minutes}m`;
}

// Helper to get status color
export function getStatusColor(status: ScanStatus): string {
  const colorMap: Record<ScanStatus, string> = {
    not_started: 'bg-gray-100 text-gray-800 border-gray-300',
    in_progress: 'bg-blue-100 text-blue-800 border-blue-300',
    paused: 'bg-yellow-100 text-yellow-800 border-yellow-300',
    completed: 'bg-green-100 text-green-800 border-green-300',
    failed: 'bg-red-100 text-red-800 border-red-300',
  };
  return colorMap[status] || 'bg-gray-100 text-gray-800 border-gray-300';
}

// Helper to get status label
export function getStatusLabel(status: ScanStatus): string {
  const labelMap: Record<ScanStatus, string> = {
    not_started: 'Not Started',
    in_progress: 'In Progress',
    paused: 'Paused',
    completed: 'Completed',
    failed: 'Failed',
  };
  return labelMap[status] || status;
}

// Helper to format ETA
export function formatETA(estimatedCompletion: string | null): string {
  if (!estimatedCompletion) return 'Calculating...';

  const eta = new Date(estimatedCompletion);
  const now = new Date();
  const diffMs = eta.getTime() - now.getTime();

  if (diffMs <= 0) return 'Soon';

  const diffMinutes = Math.floor(diffMs / 1000 / 60);
  if (diffMinutes < 1) return '< 1 min';
  if (diffMinutes < 60) return `~${diffMinutes} min`;

  const diffHours = Math.floor(diffMinutes / 60);
  return `~${diffHours}h ${diffMinutes % 60}m`;
}

File: agent_platform/api/routes/history_scan.py
"""
API Routes for History Scan operations
"""

from fastapi import APIRouter, HTTPException, Depends, BackgroundTasks
from typing import List, Optional

from agent_platform.history_scan import (
    HistoryScanService,
    ScanConfig,
    ScanProgress,
    ScanStatus,
)
from agent_platform.core.logger import get_logger

logger = get_logger(__name__)

router = APIRouter(prefix="/api/v1/history-scan", tags=["history-scan"])

# Singleton service instance (in production, use dependency injection)
_scan_service: Optional[HistoryScanService] = None


def get_scan_service() -> HistoryScanService:
    """Get or create scan service instance"""
    global _scan_service
    if _scan_service is None:
        _scan_service = HistoryScanService()
    return _scan_service


@router.post("/start", response_model=ScanProgress)
async def start_scan(
    config: ScanConfig,
    background_tasks: BackgroundTasks,
    scan_service: HistoryScanService = Depends(get_scan_service),
):
    """
    Start a new history scan

    Initiates a background scan of the Gmail account history with the provided configuration.

    **Parameters:**
    - account_id: Gmail account to scan
    - batch_size: Number of emails per batch (10-500, default 50)
    - max_results: Maximum emails to scan (None = all)
    - query: Gmail query filter (e.g., 'after:2023/01/01')
    - skip_already_processed: Skip emails already processed
    - process_attachments: Download attachments
    - process_threads: Generate thread summaries

    **Returns:**
    - ScanProgress object with scan_id and initial status
    """
    try:
        # Note: In a real implementation, we'd need to pass the gmail_service
        # For now, we return a mock response. Full integration requires Gmail OAuth.
        logger.info(f"Starting scan for account {config.account_id}")

        # In production, get gmail_service from auth system
        # gmail_service = get_gmail_service(config.account_id)
        # progress = await scan_service.start_scan(gmail_service, config)

        # For now, return mock progress
        from agent_platform.history_scan.models import ScanProgress
        import uuid
        from datetime import datetime

        progress = ScanProgress(
            scan_id=str(uuid.uuid4()),
            account_id=config.account_id,
            status=ScanStatus.IN_PROGRESS,
            total_found=0,
            processed=0,
            started_at=datetime.now(),
            last_updated_at=datetime.now(),
        )

        # Store in service
        scan_service._active_scans[progress.scan_id] = progress

        return progress

    except Exception as e:
        logger.error(f"Failed to start scan: {e}", exc_info=True)
        raise HTTPException(status_code=500, detail=f"Failed to start scan: {str(e)}")


@router.get("/{scan_id}", response_model=ScanProgress)
async def get_scan_progress(
    scan_id: str,
    scan_service: HistoryScanService = Depends(get_scan_service),
):
    """
    Get progress of a specific scan

    **Parameters:**
    - scan_id: Unique scan identifier

    **Returns:**
    - ScanProgress with current status and statistics
    """
    progress = scan_service.get_scan_progress(scan_id)
    if not progress:
        raise HTTPException(status_code=404, detail=f"Scan {scan_id} not found")

    return progress


@router.get("/", response_model=List[ScanProgress])
async def list_active_scans(
    scan_service: HistoryScanService = Depends(get_scan_service),
):
    """
    List all active scans

    **Returns:**
    - List of ScanProgress objects for all active/paused scans
    """
    return scan_service.list_active_scans()


@router.post("/{scan_id}/pause")
async def pause_scan(
    scan_id: str,
    scan_service: HistoryScanService = Depends(get_scan_service),
):
    """
    Pause an active scan

    **Parameters:**
    - scan_id: Unique scan identifier

    **Returns:**
    - Success message
    """
    success = await scan_service.pause_scan(scan_id)
    if not success:
        raise HTTPException(
            status_code=400,
            detail=f"Cannot pause scan {scan_id} (not found or not in progress)",
        )

    return {"status": "paused", "scan_id": scan_id}


@router.post("/{scan_id}/resume")
async def resume_scan(
    scan_id: str,
    scan_service: HistoryScanService = Depends(get_scan_service),
):
    """
    Resume a paused scan

    **Parameters:**
    - scan_id: Unique scan identifier

    **Returns:**
    - Success message

    **Note:** Requires gmail_service to be passed (not implemented in mock)
    """
    # In production: gmail_service = get_gmail_service(account_id)
    # success = await scan_service.resume_scan(scan_id, gmail_service)

    # For now, mock implementation
    progress = scan_service.get_scan_progress(scan_id)
    if not progress or progress.status != ScanStatus.PAUSED:
        raise HTTPException(
            status_code=400,
            detail=f"Cannot resume scan {scan_id} (not found or not paused)",
        )

    progress.status = ScanStatus.IN_PROGRESS
    return {"status": "resumed", "scan_id": scan_id}


@router.post("/{scan_id}/cancel")
async def cancel_scan(
    scan_id: str,
    scan_service: HistoryScanService = Depends(get_scan_service),
):
    """
    Cancel an active or paused scan

    **Parameters:**
    - scan_id: Unique scan identifier

    **Returns:**
    - Success message
    """
    success = await scan_service.cancel_scan(scan_id)
    if not success:
        raise HTTPException(status_code=404, detail=f"Scan {scan_id} not found")

    return {"status": "cancelled", "scan_id": scan_id}


@router.get("/{scan_id}/stats")
async def get_scan_stats(
    scan_id: str,
    scan_service: HistoryScanService = Depends(get_scan_service),
):
    """
    Get detailed statistics for a scan

    **Parameters:**
    - scan_id: Unique scan identifier

    **Returns:**
    - Detailed statistics including breakdowns by category, confidence, etc.
    """
    progress = scan_service.get_scan_progress(scan_id)
    if not progress:
        raise HTTPException(status_code=404, detail=f"Scan {scan_id} not found")

    return {
        "scan_id": scan_id,
        "account_id": progress.account_id,
        "status": progress.status,
        "progress": {
            "total_found": progress.total_found,
            "processed": progress.processed,
            "skipped": progress.skipped,
            "failed": progress.failed,
            "percent": progress.progress_percent,
        },
        "classification": {
            "high": progress.classified_high,
            "medium": progress.classified_medium,
            "low": progress.classified_low,
        },
        "extraction": {
            "tasks": progress.tasks_extracted,
            "decisions": progress.decisions_extracted,
            "questions": progress.questions_extracted,
        },
        "resources": {
            "attachments": progress.attachments_downloaded,
            "threads": progress.threads_summarized,
        },
        "timing": {
            "started_at": progress.started_at,
            "last_updated_at": progress.last_updated_at,
            "completed_at": progress.completed_at,
            "estimated_completion": progress.estimated_completion,
        },
        "error": {
            "message": progress.error_message,
            "details": progress.error_details,
        } if progress.error_message else None,
    }
